diff --git a/src/libfibre/EventScope.h b/src/libfibre/EventScope.h
index 81c600e..ea2d76b 100644
--- a/src/libfibre/EventScope.h
+++ b/src/libfibre/EventScope.h
@@ -163,13 +163,18 @@ class EventScope {
   T syncIO( T (*iofunc)(int, Args...), int fd, Args... a) {
     T ret;
     static const Poller::Direction direction = Input ? Poller::Input : Poller::Output;
-#if TESTING_EVENTPOLL_ONESHOT
+#if TESTING_EVENTPOLL_ONESHOT || TESTING_EVENTPOLL_ONDEMAND
     static const Poller::Variant variant = Input ? Poller::Oneshot : Poller::Oneshot;
 #elif TESTING_EVENTPOLL_LEVEL
     static const Poller::Variant variant = Input ? Poller::Level : Poller::Oneshot;
 #else
     static const Poller::Variant variant = Input ? Poller::Edge : Poller::Oneshot;
 #endif
+#if TESTING_EVENTPOLL_ONDEMAND
+    static const Poller::Operation pollerCreate = Input ? Poller::CreateOndemand : Poller::Create;
+#else
+    static const Poller::Operation pollerCreate = Poller::Create;
+#endif
 #if TESTING_IO_TRYFIRST
     if (Yield) Fibre::yield();
     if (tryIO<Input>(ret, iofunc, fd, a...)) return ret;
@@ -179,15 +184,15 @@ class EventScope {
     BasePoller*& poller = fdSyncVector[fd].poller[Input];
     if (!poller) {
       poller = &getPoller<Input,Cluster>(fd);
-      poller->setupFD(fd, Poller::Create, direction, variant);
-    } else if (variant == Poller::Oneshot) {
+      poller->setupFD(fd, pollerCreate, direction, variant);
+    } else if (variant == Poller::Oneshot && pollerCreate != Poller::CreateOndemand) {
       poller->setupFD(fd, Poller::Modify, direction, variant);
     }
     Poller::SyncSem& sync = fdSyncVector[fd].sync[Input];
     for (;;) {
       if (variant == Poller::Level) sync.wait(); else sync.P();
       if (tryIO<Input>(ret, iofunc, fd, a...)) return ret;
-      if (variant == Poller::Oneshot) {
+      if (variant == Poller::Oneshot && pollerCreate != Poller::CreateOndemand) {
         poller->setupFD(fd, Poller::Modify, direction, variant);
       }
     }
diff --git a/src/libfibre/Poller.h b/src/libfibre/Poller.h
index fa260ca..3b1de3a 100644
--- a/src/libfibre/Poller.h
+++ b/src/libfibre/Poller.h
@@ -44,7 +44,7 @@ struct Poller {
   enum Variant   : ssize_t { Level = 0, Edge = EV_CLEAR, Oneshot = EV_ONESHOT };
 #else // __linux__ below
   typedef epoll_event   EventType; // man 2 epoll_ctl: EPOLLERR, EPOLLHUP not needed
-  enum Operation : ssize_t { Create = EPOLL_CTL_ADD, Modify = EPOLL_CTL_MOD, Remove = EPOLL_CTL_DEL };
+  enum Operation : ssize_t { CreateOndemand = EPOLL_CTL_ADD | EPOLL_CTL_ONDEMAND, Create = EPOLL_CTL_ADD, Modify = EPOLL_CTL_MOD, Remove = EPOLL_CTL_DEL };
   enum Direction : ssize_t { Input = EPOLLIN | EPOLLPRI | EPOLLRDHUP, Output = EPOLLOUT };
   enum Variant   : ssize_t { Level = 0, Edge = EPOLLET, Oneshot = EPOLLONESHOT };
 #endif
diff --git a/src/runtime-glue/testoptions.h.default b/src/runtime-glue/testoptions.h.default
index 4bb7c3e..0a8a3ba 100644
--- a/src/runtime-glue/testoptions.h.default
+++ b/src/runtime-glue/testoptions.h.default
@@ -6,7 +6,9 @@
 #define TESTING_CLUSTER_POLLER_FIBRE  1 // per-cluster poller(s): fibre vs. pthread
 #define TESTING_CLUSTER_POLLER_FLOAT  1 // per-cluster poller fibres(s): float vs. background
 #define TESTING_IO_TRYFIRST           1 // try nonblocking I/O operation first
-#define TESTING_EVENTPOLL_ONESHOT     1 // use oneshot event polling
+#define TESTING_EVENTPOLL_ONDEMAND    1 // use ondemand event polling
+#define EPOLL_CTL_ONDEMAND            2048
+//#define TESTING_EVENTPOLL_ONESHOT     1 // use oneshot event polling
 //#define TESTING_EVENTPOLL_LEVEL       1 // use level-trigger event polling
 //#define TESTING_POLLER_FIBRE_SPIN 65536 // poller fibre: spin loop of NB polls
 
